### ISSUE TITLE
RFC: Makes salt minion's master_port and publish_port options depend on each master's ip

### ISSUE THREAD

**Issue by kstreee** [saltstack/salt#50412]
- Feature Name: Makes salt minion's master_port and publish_port options depend on each master's ip.
- Start Date: 2018-11-07
- RFC PR: 
- Salt Issue: 

# Summary
[summary]: #summary

For example, current salt minion configuration can't cover the following setting of multi-master.

master1's config
```
 interface: 1.1.1.1
 ret_port: 30001 
 publish_port: 30002
```
master2's config
```
 interface: 1.1.1.2
 ret_port: 30008
 publish_port: 30009
```

minion's config
```
master:
  1.1.1.1:30001
  1.1.1.2:30001
publish_port: ?
```

# Motivation
[motivation]: #motivation

We are trying to deploy salt master using kubernetes, and kubernetes-like systems allocate network resources highly dynamically to use the limited resources efficiently. To leverage the system's network environment (dynamically allocate ports to expose container's ports to the outer network), `publish_port` and `ret_port` of salt masters of a multi-master environment can be assigned as different ports like the above setting.

# Design
[design]: #detailed-design

Current salt minion's master config
```
master:
  1.1.1.1:RET_PORT (MASTER_PORT)
  1.1.1.2:RET_PORT (MASTER_PORT)
publish_port: PUBLISH_PORT
```

Desired salt minion's master config
```
master:
  - host: 1.1.1.1
    master_port: 30001
    publish_port: 30002
  - host: 1.1.1.2
    master_port: 30008
    publish_port: 30009
```

## Alternatives
[alternatives]: #alternatives

Giving up multi-master setting which uses different ports for the same purpose..?

## Unresolved questions
[unresolved]: #unresolved-questions

Not sure... how to find all being affected points in the minion code?

# Drawbacks
[drawbacks]: #drawbacks

- Backward compatibility, how to support both formats (old format, and new format) of master connection information of salt minion config?
- Implementation cost, we should find all being affected points in the minion code, and patch it to accept the new form of salt master connection config.
- Documentation

**Comment by dwoz** [comment]
@kstreee Thanks for the feature request. It makes sense to me and I like it.. @core-team comments?

**Comment by isbm** [comment]
What's wrong would be just adding the `publish_port` instead?
```yaml
master:
  - [2001:0db8:85a3:0000:0000:8a2e:0370:7334]:3001
    publish_port: 3002
  - 123.123.123.123:3001
    publish_port: 3002
  - foo.bar.blah.com:3001
    publish_port: 3002
  - publish_port: 3002  # Default port
```

**Comment by isbm** [comment]
Second, this have to be an RFC pull request, not an issue. Please close the issue and rewrite this as an RFC, and make a PR so we can discuss it there.

**Comment by kstreee** [comment]
Sure, I would rather modify what I sent as a RFC. When I opened this issue, I thought to add this feature might be tricky so I opened this issue to discuss before providing sample code of this RFC. But, I realized it isn't after dived into the code, so to show a sample and modification points of this RFC, sent a PR. As your advice, closed this issue. And moved your comment to the pr here #50521. Thanks for the guiding me.
